misk/src/main/kotlin/misk/digester/FakeDigest.kt

```diff
@@ -0,0 +1,61 @@
+package misk.digester
+
+import com.squareup.digester.protos.service.DigestData
+
+/**
+ * Wraps an adapted t-digest implementation from Stripe's Veneur project
+ */
+class FakeDigest : TDigest {
+  constructor() {
+    addedValues = mutableListOf()
+  }
+
+  constructor (values: List<Double>) {
+    addedValues = values.toMutableList()
+  }
+
+  internal val addedValues: MutableList<Double>
+  private var count: Long = 0
+  private var sum: Double = 0.0
+  /** Adds a new observation to the t-digest */
+  override fun add(value: Double) {
+    addedValues.add(value)
+    count += 1
+    sum += value
+  }
+
+  override fun quantile(quantile: Double): Double {
+    if (addedValues.count() > 0) {
+      return addedValues.last()
+    }
+    return Double.NaN
+  }
+
+  /** Returns the count of the number of observations recorded within the t-digest */
+  override fun count(): Long {
+    return count
+  }
+
+  /** Returns the sum of all values added into the digest, or NaN if no values have been added */
+  override fun sum(): Double {
+    if (count > 0) {
+      return sum
+    }
+    return Double.NaN
+  }
+
+  /** Merges this t-digest into another t-digest */
+  override fun mergeInto(other: TDigest) {
+    require(other is FakeDigest) {
```

Comment:
Is there a better way of overriding the TDigest mergeInto function to accept only FakeDigest? Maybe through guice?

Reply:
You can use the type system. Here's an example in Java (apologies for not being able to write Kotlin yet).

```
import java.io.*;
import java.util.*;

/*
 * To execute Java, please define "static void main" on a class
 * named Solution.
 *
 * If you need more classes, simply define them inline.
 */

class Solution {
  static interface TDigest<T extends TDigest<T>> {
    void mergeInto(T other);
  }
  
  static class FakeTDigest implements TDigest<FakeTDigest> {
    public void mergeInto(FakeTDigest other) {}
  }
  
  static class RealTDigest implements TDigest<RealTDigest> {
    public void mergeInto(RealTDigest other) {}
  }
  
  public static void main(String[] args) {
    FakeTDigest fake = new FakeTDigest();
    RealTDigest real = new RealTDigest();
    
    real.mergeInto(real);
    // real.mergeInto(fake); // won't compile
  }
}
```